ir/memory.cpp

```diff
@@ -218,17 +223,27 @@ pair<expr, vector<expr>> Memory::mkInput(const char *name) {
   return { Pointer(*this, offset, local_bid, bid).release(), { var } };
 }
 
-expr Memory::alloc(const expr &bytes, unsigned align, bool local) {
+expr Memory::alloc(const expr &bytes, unsigned align, bool local, bool heap) {
   Pointer p(*this, ++last_bid, local);
   state->addPre(p.is_aligned(align));
 
   expr size = bytes.zextOrTrunc(bits_size_t);
   state->addPre(p.block_size() == size);
+
+  state->addPre(blocks_liveness.load(p.get_bid()) == 0);
```

Comment:
We should we use the initial `blocks_liveness` for this.  Otherwise code following another branch may conclude that the block was live before.

Reply:
I think it is still okay because it still satisfies uniqueness of block id. It would depend on how the join operation on memory is defined as well.

Reply:
Oh, does your comment mean this precondition may not hold with any bid, leading to making Z3 always answer UNSAT?
If my understanding is correct, it would be related with Memory::bits_for_local/nonlocal_bid which is the number of bits that are used to encode bids? If Memory::bits_for_local/nonlocal_bid are large enough, Z3 will be able to assign proper bids so so this precondition can hold.

Reply:
Imagine a diamond CFG;
```
      A
    /   \
   B     C
    \   /
      D
```

If an allocation happens in basic block B, the block id gets reserved there and we say that it is now live. But what happens before in block A? And in block C?  And when you merge in D?
You need to state that the block wasn't alive in A. So you can't use `blocks_liveness` of block B. You need to state the property w.r.t. to the original array before execution of A even.

Reply:
> You need to state that the block wasn't alive in A.

`blocks_liveness` is preserved before/after execution of instructions that are not malloc/free/alloca, as Memory's `blocks_val` is preserved after execution of non-memory instructions.
If A has no malloc/free/alloca, `blocks_liveness` before A will be equivalent to that of B.
So, this precondition snippet is equivalent to stating that the block wasn't alive before A.

The underlying principle of `blocks_liveness` is that `malloc` is equivalent to a `store`  that writes bit 1 to a hidden memory that records liveness, after checking that originally it was 0.
Similarly, `free` is equivalent to another `store` that writes 0 after checking 1 was written (otherwise UB due to double-free).
So, I believe it is enough to see the memory just before execution of this malloc instruction to avoid having two live blocks with the same id.

Reply:
The issue is when there's nothing written in the liveness record. That means that the block may be alive or not. When you reserve a block id you need to assert that it was dead in the beginning of the function and becomes alive after the malloc. But it couldn't haven't been possibly alive before.
The current code doesn't implement this and is therefore incorrect. It's also creating a bigger expression than needed because `blocks_liveness` at the current BB will already have plenty of stuff in there.